tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/ant/GenerateCodeTask.java

```diff
@@ -0,0 +1,93 @@
+import org.apache.tools.ant.BuildException;
+import org.apache.tools.ant.Task;
+
+import java.io.*;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+// Note: Auto-magical Ant methods are listed as "unused" by IDEs, unless the warning is suppressed.
+public final class GenerateCodeTask extends Task {
+    private Path cldrPath;
+    private File cOutDir;
+    private File javaOutDir;
+    private String action;
+
+    private class GeneratedFileDef {
+        String cRelativePath;
+        String javaRelativePath;
+        CodeGenerator generator;
+
+        public GeneratedFileDef(String cRelativePath, String javaRelativePath, CodeGenerator generator) {
+            this.cRelativePath = cRelativePath;
+            this.javaRelativePath = javaRelativePath;
+            this.generator = generator;
+        }
+    }
+
+    private GeneratedFileDef[] generatedFileDefs = {
+        new GeneratedFileDef("common/localeFallbackTables.h", "src/com/ibm/icu/impl/LocaleFallbackTables.java", new ResourceFallbackCodeGenerator()),
+    };
+
+    @SuppressWarnings("unused")
+    public void setCldrDir(String path) {
+        // Use String here since on some systems Ant doesn't support automatically converting Path instances.
+        this.cldrPath = checkNotNull(Paths.get(path));
+    }
+
+    @SuppressWarnings("unused")
+    public void setCOutDir(String path) {
+        // Use String here since on some systems Ant doesn't support automatically converting Path instances.
+        this.cOutDir = new File(path);
+    }
+
+    @SuppressWarnings("unused")
+    public void setJavaOutDir(String path) {
+        // Use String here since on some systems Ant doesn't support automatically converting Path instances.
+        this.javaOutDir = new File(path);
+    }
+
+    @SuppressWarnings("unused")
+    public void setAction(String action) {
+        // Use String here since on some systems Ant doesn't support automatically converting Path instances.
+        this.action = action;
+    }
+
+    @SuppressWarnings("unused")
+    public void execute() throws BuildException {
+        for (GeneratedFileDef task : generatedFileDefs) {
+            File cOutPath = new File(cOutDir, task.cRelativePath);
+            File javaOutPath = new File(javaOutDir, task.javaRelativePath);
+
+            if (this.action != null && this.action.equals("clean")) {
+                cOutPath.delete();
+                javaOutPath.delete();
+            } else {
+                log("Deleting " + cOutPath + " and " + javaOutPath + "...");
+                cOutPath.getParentFile().mkdirs();
+                javaOutPath.getParentFile().mkdirs();
+
+                try {
+                    PrintWriter cOut = new PrintWriter(new OutputStreamWriter(new FileOutputStream(cOutPath)));
+                    PrintWriter javaOut = new PrintWriter(new OutputStreamWriter(new FileOutputStream(javaOutPath)));
```

Comment:
Could you please use try-with-resources?
- https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html
- https://www.baeldung.com/java-try-with-resources

Reply:
Oh, that's slick.  I didn't know that existed either.  Fixed.